### DISCUSSION TITLE
how do you count pulses?

### DISCUSSION THREAD

**Discussion by jbeale1** [micropython/micropython#13053] [ESP32]
I want to count some pulses, and they might come more quickly than handling them in an interpreted language probably allows, so I'd like to use hardware. I'm curious if anyone ever counts pulses in micropython now, and if so how do they do it? I've seen some discussion on other hardware (using the PIO module with Pi Pico). While I know the ESP32 has a hardware module for this (PCNT) it seems there is no support for that yet in micropython, although it has been discussed at least since 2019, and separately a quadrature encoder since 2022. https://github.com/micropython/micropython/issues/5216 https://github.com/micropython/micropython/pull/8766

**Comment by robert-hh** [comment]
There are a few PR's outstanding for supporting counter and encoder functionality. And @jimmo mentioned that there is work going on to support it. The discussion is still about the API and for ESP32 the best way to support it. That said, I cannto tell how long it will take to have in in main line. Until then, you could build  your own firmware based on the existing PRs for the ESP32 or MIMXRT platform.

**Reply by paukstelis** [reply]
Could someone point me towards semi-recent PRs that include the pulse counter? It seems like everything I've come across is quite old.

**Comment by kjm1102** [comment]
You can count pulses in your sleep https://github.com/micropython/micropython-esp32-ulp if you want to

**Comment by robert-hh** [comment]
For a PR it is less important when it was made, but whether it is maintained and updated along the progress of the main branch. As long as this happens, it is recent. #8766 was updated 2 months ago. Maybe @IhorNehrutsa can update it now again.  There is as well PR #12346 intended to replace #8766 and #7582, which is newer but seems to require maintenance by it's author @jimmo. I do not know if there will by any progress of this feature. Maybe it's going just to another hibernation period.

**Reply by paukstelis** [reply]
Thanks for the info, and thanks to @IhorNehrutsa for the nearly immediate update!

**Comment by MRIIOT** [comment]
How fast, IRQ seems to work fine up to 10kHz in my tests.

```
gc.disable()

count = 0

def handle_irq(pin):
	global count
	count += 1

pin_id = 39
p = Pin(pin_id, Pin.IN, Pin.PULL_UP)
p.irq(trigger=Pin.IRQ_RISING, handler=handle_irq) 

print('start loop')
while True:
	time.sleep_ms(1000)
	print('t={} c={}'.format(time.ticks_ms(), count))
	state = machine.disable_irq()
	count = 0
	machine.enable_irq(state)
```

**Reply by samneggs** [reply]
Interesting. With Viper I was able to get to 40kHz.
And when overclocked to 200MHz I got to around 65kHz.
```
ARY = array.array('I',[0])
@micropython.viper
def handle_irq(pin):
    ary = ptr32(ARY)
    ary[0] += 1
